File: web/backend/models/cv.py
from pydantic import BaseModel, EmailStr, Field
from typing import List, Optional
from datetime import datetime


class Education(BaseModel):
    institution: str
    degree: str
    field_of_study: str
    start_date: str
    end_date: Optional[str] = None
    description: Optional[str] = None
    location: Optional[str] = None
    achievements: Optional[List[str]] = None


class Experience(BaseModel):
    company: str
    position: str
    start_date: str
    end_date: Optional[str] = None
    description: Optional[str] = None
    location: Optional[str] = None
    achievements: Optional[List[str]] = None
    technologies: Optional[List[str]] = None


class Skill(BaseModel):
    name: str
    level: Optional[int] = Field(None, ge=1, le=5)
    category: Optional[str] = None
    years_of_experience: Optional[int] = None


class Certificate(BaseModel):
    name: str
    issuer: str
    date_obtained: str
    expiry_date: Optional[str] = None
    credential_id: Optional[str] = None
    credential_url: Optional[str] = None


class Language(BaseModel):
    name: str
    level: str
    certification: Optional[str] = None


class CV(BaseModel):
    full_name: str
    email: EmailStr
    phone: Optional[str] = None
    summary: Optional[str] = None
    education: List[Education] = []
    experience: List[Experience] = []
    skills: List[Skill] = []
    certificates: Optional[List[Certificate]] = []
    languages: Optional[List[Language]] = []
    website: Optional[str] = None
    linkedin: Optional[str] = None
    github: Optional[str] = None
    photo_url: Optional[str] = None
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None
    version: Optional[int] = 1

File: web/frontend/src/types/cv.ts
export interface Education {
  institution: string;
  degree: string;
  fieldOfStudy: string;
  startDate: string;
  endDate?: string;
  description?: string;
  location?: string;
  achievements?: string[];
}

export interface Experience {
  company: string;
  position: string;
  startDate: string;
  endDate?: string;
  description?: string;
  location?: string;
  achievements?: string[];
  technologies?: string[];
}

export interface Skill {
  name: string;
  level?: number;
  category?: string;
  yearsOfExperience?: number;
}

export interface Certificate {
  name: string;
  issuer: string;
  dateObtained: string;
  expiryDate?: string;
  credentialId?: string;
  credentialUrl?: string;
}

export interface Language {
  name: string;
  level: string;
  certification?: string;
}

export interface CVData {
  fullName: string;
  email: string;
  phone?: string;
  summary?: string;
  education: Education[];
  experience: Experience[];
  skills: Skill[];
  certificates: Certificate[];
  languages: Language[];
  website?: string;
  linkedin?: string;
  github?: string;
  photoUrl?: string;
}

export type CVDataApi = {
  full_name: string;
  email: string;
  phone?: string;
  summary?: string;
  education: {
    institution: string;
    degree: string;
    field_of_study: string;
    start_date: string;
    end_date?: string;
    description?: string;
    location?: string;
    achievements?: string[];
  }[];
  experience: {
    company: string;
    position: string;
    start_date: string;
    end_date?: string;
    description?: string;
    location?: string;
    achievements?: string[];
    technologies?: string[];
  }[];
  skills: {
    name: string;
    level?: number;
    category?: string;
    years_of_experience?: number;
  }[];
  certificates: {
    name: string;
    issuer: string;
    date_obtained: string;
    expiry_date?: string;
    credential_id?: string;
    credential_url?: string;
  }[];
  languages: {
    name: string;
    level: string;
    certification?: string;
  }[];
  website?: string;
  linkedin?: string;
  github?: string;
  photo_url?: string;
};

function toSnakeCase(str: string): string {
  return str.replace(/[A-Z]/g, (letter) => `_${letter.toLowerCase()}`);
}

function transformValue(val: unknown): unknown {
  if (Array.isArray(val)) {
    return val.map((item) =>
      item && typeof item === 'object' ? transformKeysToSnake(item as Record<string, unknown>) : item,
    );
  }
  return val;
}

function transformKeysToSnake(obj: Record<string, unknown>): Record<string, unknown> {
  const result: Record<string, unknown> = {};
  for (const [key, value] of Object.entries(obj)) {
    result[toSnakeCase(key)] = transformValue(value);
  }
  return result;
}

export function toApiFormat(data: CVData): CVDataApi {
  return transformKeysToSnake(data as unknown as Record<string, unknown>) as unknown as CVDataApi;
}